handout/testing/Unit Testing Components.md

```diff
@@ -0,0 +1,240 @@
+  return tcb.createAsync(QuoteComponent).then(fixture) => {
+  		fixture.detectChanges();
+    	fixture.debugElement.componentInstance.getQuote();
+    	
+    	fixture.detectChanges();
+	    var compiled = fixture.debugElement.nativeElement;
+    	expect(compiled.querySelector('div')).toHaveText('Test Quote');	
+	}
+  }));
+});
+```
+
+Testing the QuoteComponent is a fairly straightforward process. We want to create a QuoteComponent, feed it a quote and see if the DOM has changed such that the quote has shown up. This process requires us to create the component, pass in any dependencies, trigger the component to perform an action, and then look at the DOM to see if the action is what we expected. Lets take a look at how this is accomplished with the above unit test. 
+
+We use `beforeEachProviders` to feed in any dependencies that our component requires. Here our component depends on the `QuoteService` to get data, we mock this data ourselves as we can have control over what value we expect to show up. It is good practise to seperate component testing from service testing - this makes it easier to test as you are only focusing on a single aspect of the application at a time. If your service fails, or your component fails, how will you know which one was the culprit? Using `provide` we inject the `QuoteService` dependency using our mock class `MockQuoteService` , where we will provide mock data for the component to consume. 
+
+Next we use `injectAsync` to inject the `TestComponentBuilder` into our test. Once we have a reference to `TestComponentBuilder` we call `createAsync` to create the component we will be testing - `QuoteComponent`. The `TestComponentBuilder` will then create a new instance of our component, fufilling any Angular specific routines like dependency injection. The `createAsync` returns a promise, which we return as `injectAsync` expects a promise to know when the test is done. 
+
+`TestComponentBuilder` will return a *fixture* for us to use in our tests. A fixture is powerful tool athat allows us to query the DOM rendered by a component as well as change DOM elements and component properties. It is the main point of testing components and we use it extensively. Here we have gotten access to our component through the `fixture.debugElement.componentInstance` property where we call `getQuote` to kickstart our only action in the `QuoteComponent` component. We call `fixture.detectChanges` to watch for any changes in the DOM. We then get access to the native underlying DOM elements through the `fixture.debugElement.nativeElement` property. Now we can check to see if the DOM rendered by our `QuoteComponent` contains the quote that we mocked in through the `QuoteService`. The final line, attempts to assert that the DOM's div tag contains the mocked quote 'Test Qoute' inside. If it does, then our component passes the test and works as expected, if it doesn't, that means our component is not outputting quotes correctly. 
```

Comment:
"We call `fixture.detectChanges` to watch for any changes in the DOM". Maybe we should avoid the word **watch** because it has a specific connotation in Angular 1. How about just saying "We call `fixture.detectChanges` to update the DOM with the new value of `this.quote`"?